wikidict/lang/fr/template_handlers.py

```diff
@@ -207,72 +208,98 @@ def render_compose_de(tpl: str, parts: List[str], data: Dict[str, str]) -> str:
     'Composé de <i>faire</i>, <i>boutique</i> et <i>cul</i>'
     >>> render_compose_de("composé de", ["arthro-", "-logie"], defaultdict(str, {"lang": "fr", "m": "oui"}))
     'Composé de <i>arthro-</i> et de <i>-logie</i>'
+    >>> render_compose_de("composé de", ["morin", "morine", "-elle"], defaultdict(str, {"lang": "fr", "m": "1"}))
+    'Composé de <i>morin</i>, <i>morine</i> et <i>-elle</i>'
+    >>> render_compose_de("composé de", ["bi-", "mensis"], defaultdict(str, {"lang": "fr", "sens1": "deux", "sens2":"mois"}))
+    'dérivé de <i>mensis</i> («&nbsp;mois&nbsp;») avec le préfixe <i>bi-</i> («&nbsp;deux&nbsp;»)'
+    >>> render_compose_de("composé de", ["im-", "brouiller", "-able"], defaultdict(str, {"lang": "fr", "m": "oui"}))
+    'Dérivé de <i>brouiller</i> avec le préfixe <i>im-</i> et le suffixe <i>-able</i>'
+    >>> render_compose_de("composé de", ["bloc", "d’", "obturation", "de", "puits"], defaultdict(str, {"lang": "fr", "m": "1", "f": "1"}))
+    'Composée de <i>bloc</i>, <i>d’</i>, <i>obturation</i>, <i>de</i> et <i>puits</i>'
+    >>> render_compose_de("composé de", ["an-", "", "-onyme"], defaultdict(str, {"lang": "fr", "m": "1"}))
+    'Dérivé du préfixe <i>an-</i> et le suffixe <i>-onyme</i>'
+    >>> render_compose_de("composé de", ["an-"], defaultdict(str))
+    'dérivé du préfixe <i>an-</i>'
     """  # noqa
-    is_not_derived = all(part.startswith("-") or part.endswith("-") for part in parts)
-    is_derived = not is_not_derived and any(
-        part.startswith("-") or part.endswith("-") for part in parts
-    )
-    is_derived |= any(
-        part.startswith("-") or part.endswith("-") for part in data.values()
-    )
+
+    # algorithm from https://fr.wiktionary.org/w/index.php?title=Mod%C3%A8le:compos%C3%A9_de&action=edit
+    p1 = data.get("tr1", "") or parts[0] if parts else ""
+    b1 = "1" if p1.endswith("-") else "0"
+    p2 = data.get("tr2", "") or parts[1] if len(parts) > 1 else ""
+    b2 = "1" if p2.startswith("-") else "0"
+    b3 = "0"
+    if len(parts) > 2:
+        p3 = data.get("tr3", "") or parts[2] if len(parts) > 2 else ""
+        b3 = "2" if p3.startswith("-") else "1"
+    b4 = "1" if len(parts) > 3 else "0"
+
+    b = b1 + b2 + b3 + b4
+    is_derived = b == "1000" or b == "0100" or b == "1020"
+
+    def word_tr_sens(w: str, tr: str, sens: str) -> str:
```

Comment:
This function could be out of the `render_compose_de` function and be reused in other templates such as etyl.
However, it's a hack regarding the italic handling. The right way to do it it's to check if each char of the words are "Latin", "Common" or "Inherited" in the script ranges from https://fr.wiktionary.org/wiki/Module:donn%C3%A9es_Unicode/data/script_ranges

I tried another way in render_etyl but it's not working for d’ (not the apostrophe...)

@BoboTiG WDYT ?

Reply:
We could move the function and handle unicode chars correctly, yes. It could be done in another PR to keep this one as-is (already quite complex).